File: packages/metadata/python/metadata/schema.py
"""Pydantic v2 schemas for the Domain Explorer metadata registry.

The shapes mirror packages/metadata/src/schema.ts (Zod) so the same YAML
files validate identically in both ecosystems.
"""
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class Vertical(str, Enum):
    BFSI = "BFSI"
    INSURANCE = "Insurance"
    RETAIL = "Retail"
    RCG = "RCG"
    CPG = "CPG"
    TTH = "TTH"
    MANUFACTURING = "Manufacturing"
    LIFE_SCIENCES = "LifeSciences"
    HEALTHCARE = "Healthcare"
    TELECOM = "Telecom"
    MEDIA = "Media"
    ENERGY = "Energy"
    UTILITIES = "Utilities"
    PUBLIC_SECTOR = "PublicSector"
    HI_TECH = "HiTech"
    PROFESSIONAL_SERVICES = "ProfessionalServices"


KpiDirection = Literal["higher_is_better", "lower_is_better", "target_band"]
PersonaLevel = Literal["C-suite", "VP", "Director", "Manager", "IC"]
ConnectorLatency = Literal["realtime", "near-realtime", "batch"]
ConnectorMode = Literal["push", "pull", "stream", "file"]


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class Persona(_Strict):
    name: str = Field(..., min_length=1)
    title: str = Field(..., min_length=1)
    level: PersonaLevel


class Decision(_Strict):
    id: str = Field(..., min_length=1)
    statement: str = Field(..., min_length=1)


class Kpi(_Strict):
    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    formula: str = Field(..., min_length=1)
    unit: str = Field(..., min_length=1)
    direction: KpiDirection
    decisionsSupported: list[str] = Field(default_factory=list)


class Entity(_Strict):
    name: str = Field(..., min_length=1)
    description: str | None = None
    keys: list[str] = Field(default_factory=list)


class DataModel(_Strict):
    entities: list[Entity] = Field(default_factory=list)


class SourceSystem(_Strict):
    vendor: str = Field(..., min_length=1)
    product: str = Field(..., min_length=1)
    category: str = Field(..., min_length=1)


class Connector(_Strict):
    type: str = Field(..., min_length=1)
    protocol: str = Field(..., min_length=1)
    auth: str = Field(..., min_length=1)


class Subdomain(_Strict):
    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    vertical: Vertical
    oneLiner: str = Field(..., min_length=1)
    personas: list[Persona] = Field(..., min_length=1)
    decisions: list[Decision] = Field(default_factory=list)
    kpis: list[Kpi] = Field(default_factory=list)
    dataModel: DataModel = Field(default_factory=DataModel)
    sourceSystems: list[SourceSystem] = Field(default_factory=list)
    connectors: list[Connector] = Field(default_factory=list)
    ingestionChallenges: list[str] = Field(default_factory=list)
    integrationChallenges: list[str] = Field(default_factory=list)


class KpiRegistryEntry(Kpi):
    vertical: Vertical


class SourceSystemEntry(_Strict):
    id: str = Field(..., min_length=1)
    vendor: str = Field(..., min_length=1)
    product: str = Field(..., min_length=1)
    category: str = Field(..., min_length=1)
    primaryConnectors: list[str] = Field(default_factory=list)


class ConnectorPattern(_Strict):
    id: str = Field(..., min_length=1)
    type: str = Field(..., min_length=1)
    protocol: str = Field(..., min_length=1)
    auth: str = Field(..., min_length=1)
    typicalSources: list[str] = Field(default_factory=list)
    latency: ConnectorLatency
    modes: list[ConnectorMode] = Field(default_factory=list)

File: packages/metadata/src/schema.ts
import { z } from "zod";

/** Vertical / industry buckets used across the registry. */
export const Vertical = z.enum([
  "BFSI",
  "Insurance",
  "Retail",
  "RCG",
  "CPG",
  "TTH",
  "Manufacturing",
  "LifeSciences",
  "Healthcare",
  "Telecom",
  "Media",
  "Energy",
  "Utilities",
  "PublicSector",
  "HiTech",
  "ProfessionalServices",
]);
export type Vertical = z.infer<typeof Vertical>;

export const KpiDirection = z.enum(["higher_is_better", "lower_is_better", "target_band"]);
export type KpiDirection = z.infer<typeof KpiDirection>;

export const Persona = z.object({
  name: z.string().min(1),
  title: z.string().min(1),
  level: z.enum(["C-suite", "VP", "Director", "Manager", "IC"]),
});
export type Persona = z.infer<typeof Persona>;

export const Decision = z.object({
  id: z.string().min(1),
  statement: z.string().min(1),
});
export type Decision = z.infer<typeof Decision>;

export const Kpi = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  formula: z.string().min(1),
  unit: z.string().min(1),
  direction: KpiDirection,
  decisionsSupported: z.array(z.string()).default([]),
});
export type Kpi = z.infer<typeof Kpi>;

/** Single column on an entity, used by the ERD viewer. */
export const EntityAttribute = z.object({
  name: z.string().min(1),
  type: z.string().min(1),
  description: z.string().optional(),
  isPrimaryKey: z.boolean().optional(),
  isForeignKey: z.boolean().optional(),
  references: z.string().optional(),
});
export type EntityAttribute = z.infer<typeof EntityAttribute>;

/** Cardinality between two entities. */
export const RelationshipKind = z.enum(["one_to_one", "one_to_many", "many_to_many"]);
export type RelationshipKind = z.infer<typeof RelationshipKind>;

export const EntityRelationship = z.object({
  to: z.string().min(1),
  kind: RelationshipKind,
  via: z.string().optional(),
});
export type EntityRelationship = z.infer<typeof EntityRelationship>;

export const Entity = z.object({
  name: z.string().min(1),
  description: z.string().optional(),
  keys: z.array(z.string()).default([]),
  attributes: z.array(EntityAttribute).default([]),
  relationships: z.array(EntityRelationship).default([]),
});
export type Entity = z.infer<typeof Entity>;

/** Pointers to per-style DDL files (relative paths) for a subdomain. */
export const DataModelArtifacts = z.object({
  threeNF: z.string().optional(),
  vault: z.string().optional(),
  dimensional: z.string().optional(),
});
export type DataModelArtifacts = z.infer<typeof DataModelArtifacts>;

export const DataModel = z.object({
  entities: z.array(Entity).default([]),
});
export type DataModel = z.infer<typeof DataModel>;

export const SourceSystem = z.object({
  vendor: z.string().min(1),
  product: z.string().min(1),
  category: z.string().min(1),
});
export type SourceSystem = z.infer<typeof SourceSystem>;

export const Connector = z.object({
  type: z.string().min(1),
  protocol: z.string().min(1),
  auth: z.string().min(1),
});
export type Connector = z.infer<typeof Connector>;

export const Subdomain = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  vertical: Vertical,
  oneLiner: z.string().min(1),
  personas: z.array(Persona).min(1),
  decisions: z.array(Decision).default([]),
  kpis: z.array(Kpi).default([]),
  dataModel: DataModel.default({ entities: [] }),
  dataModelArtifacts: DataModelArtifacts.optional(),
  sourceSystems: z.array(SourceSystem).default([]),
  connectors: z.array(Connector).default([]),
  ingestionChallenges: z.array(
    z.union([
      z.string(),
      z.object({}).passthrough()
    ])
  ).default([]),
  integrationChallenges: z.array(
    z.union([
      z.string(),
      z.object({}).passthrough()
    ])
  ).default([]),
});
export type Subdomain = z.infer<typeof Subdomain>;

/** Standalone KPI registry entry (KPIs that aren't tied to one subdomain). */
export const KpiRegistryEntry = Kpi.extend({
  vertical: Vertical,
});
export type KpiRegistryEntry = z.infer<typeof KpiRegistryEntry>;

/** Master KPI library entry — superset of registry with extra metadata. */
export const KpiMasterEntry = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  formula: z.string().min(1),
  unit: z.string().min(1),
  direction: KpiDirection,
  definition: z.string().optional(),
  vertical: Vertical.optional(),
  subdomains: z.array(z.string()).default([]),
  related_personas: z.array(z.string()).default([]),
  decisionsSupported: z.array(z.string()).default([]),
});
export type KpiMasterEntry = z.infer<typeof KpiMasterEntry>;

/** Per-style SQL implementations of a single KPI. */
export const KpiSqlSpec = z.object({
  kpi_id: z.string().min(1),
  threeNF: z.string().optional(),
  vault: z.string().optional(),
  dimensional: z.string().optional(),
  notes: z.string().optional(),
});
export type KpiSqlSpec = z.infer<typeof KpiSqlSpec>;

/** Source system registry entry. */
export const SourceSystemEntry = z.object({
  id: z.string().min(1),
  vendor: z.string().min(1),
  product: z.string().min(1),
  category: z.string().min(1),
  primaryConnectors: z.array(z.string()).default([]),
});
export type SourceSystemEntry = z.infer<typeof SourceSystemEntry>;

/** Connector pattern entry. */
export const ConnectorPattern = z.object({
  id: z.string().min(1),
  type: z.string().min(1),
  protocol: z.string().min(1),
  auth: z.string().min(1),
  typicalSources: z.array(z.string()).default([]),
  latency: z.enum(["realtime", "near-realtime", "batch"]),
  modes: z.array(z.enum(["push", "pull", "stream", "file"])).default([]),
});
export type ConnectorPattern = z.infer<typeof ConnectorPattern>;

/** Glossary term entry. */
export const GlossaryTerm = z.object({
  name: z.string().min(1),
  aliases: z.array(z.string()).default([]),
  definition: z.string().min(1),
  related_subdomains: z.array(z.string()).default([]),
  related_kpis: z.array(z.string()).default([]),
  steward: z.string().optional(),
});
export type GlossaryTerm = z.infer<typeof GlossaryTerm>;
